# Exploración Inicial

Se realiza una primera aproximación a los datos con el objetivo de comprender su estructura general, identificar el volumen de registros, examinar las variables disponibles y sus tipos de datos, así como detectar posibles inconsistencias o valores faltantes que requieran atención antes de proceder con las etapas de preprocesamiento y análisis exploratorio.

## Carga de Librerías

Se importan las librerías necesarias para las etapas de análisis de datos. Se utiliza `pandas` y `numpy` para la manipulación y transformación de datos, junto con `IPython.display` para la visualización de resultados en el entorno interactivo. Además, se suprimen los mensajes de advertencia (`warnings`) con el fin de mantener una salida limpia y centrada en los resultados del análisis.

In [1]:
import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np 

from IPython.display import display
from IPython.display import HTML

## Carga del DataFrame

Se carga el archivo CSV que contiene los datos de comparendos electrónicos utilizando la función `read_csv` de la librería pandas, y se almacena en el DataFrame `df_comparendos_electronicos`.

In [2]:
df_comparendos_electronicos = pd.read_csv("C:/Users/david/Documents/seminario_investigativo/comparendos_electronicos.csv")

## Visualización Inicial del DataFrame

Se muestra una vista previa de las primeras filas del DataFrame `df_comparendos_electronicos` utilizando el método `head()`, lo que permite inspeccionar la estructura y los primeros registros de los datos cargados.

In [3]:
df_comparendos_electronicos.head()

,fecha_comparendo,COD_INFRACCION,DESC_INFRACCION,TIPO_INFRACCION,SERVICIO_VEHICULO_INFRACTOR,CLASE_VEHICULO_INFRACTOR,CANTIDAD_INFRACCIONES,Tipo Camara,Camara_y_direccion
0,2018 Jan 01 12:00:00 AM,C29,Conducir un vehículo a velocidad superior a la...,TRÁNSITO,PARTICULAR,AUTOMOVIL,4,Fijo,AVENIDA CIRCUNVALAR CON CARRERA 31
1,2018 Jan 01 12:00:00 AM,C29,Conducir un vehículo a velocidad superior a la...,TRÁNSITO,PARTICULAR,CAMIONETA,2,Fijo,AVENIDA CIRCUNVALAR CON CARRERA 31
2,2018 Jan 01 12:00:00 AM,C29,Conducir un vehículo a velocidad superior a la...,TRÁNSITO,PARTICULAR,CAMIONETA,1,Fijo,AVENIDA CIRCUNVALAR CON CARRERA 31
3,2018 Jan 01 12:00:00 AM,C29,Conducir un vehículo a velocidad superior a la...,TRÁNSITO,PARTICULAR,AUTOMOVIL,2,Fijo,AVENIDA CIRCUNVALAR CON CARRERA 9G
4,2018 Jan 01 12:00:00 AM,C29,Conducir un vehículo a velocidad superior a la...,TRÁNSITO,PARTICULAR,CAMIONETA,3,Fijo,AVENIDA CIRCUNVALAR CON CARRERA 9G


Se presentan las últimas filas del DataFrame `df_comparendos_electronicos` mediante el método `tail()`, permitiendo verificar los registros finales y confirmar la correcta carga completa de los datos.

In [4]:
df_comparendos_electronicos.tail()

,fecha_comparendo,COD_INFRACCION,DESC_INFRACCION,TIPO_INFRACCION,SERVICIO_VEHICULO_INFRACTOR,CLASE_VEHICULO_INFRACTOR,CANTIDAD_INFRACCIONES,Tipo Camara,Camara_y_direccion
339611,2025 Dec 31 12:00:00 AM,C29,Conducir un vehículo a velocidad superior a la...,TRÁNSITO,PARTICULAR,CAMPERO,1,Fijo,CARRERA 46 CON CALLE 100
339612,2025 Dec 31 12:00:00 AM,C29,Conducir un vehículo a velocidad superior a la...,TRÁNSITO,PARTICULAR,CAMIONETA,1,Fijo,CARRERA 53 ENTRE CALLE 104 Y 106
339613,2025 Dec 31 12:00:00 AM,D04,No detenerse ante luz roja o amarilla de semáf...,TRÁNSITO,PARTICULAR,MOTOCICLETA,2,Fijo,CALLE 45 CON CARRERA 21 SENTIDO NORTE-SUR
339614,2025 Dec 31 12:00:00 AM,D04,No detenerse ante luz roja o amarilla de semáf...,TRÁNSITO,PARTICULAR,CAMPERO,1,Fijo,CARRERA 53 ENTRE CALLE 104 Y 106
339615,2025 Dec 31 12:00:00 AM,D04,No detenerse ante luz roja o amarilla de semáf...,TRÁNSITO,PARTICULAR,MOTOCICLETA,3,Fijo,CARRERA 53 ENTRE CALLE 104 Y 106


## Información General del DataFrame

Se utiliza el método `info()` para obtener un resumen de la estructura del DataFrame, incluyendo el número de entradas, los nombres de las columnas, los tipos de datos y la cantidad de valores no nulos, lo que facilita la identificación de posibles valores faltantes y la verificación de la integridad de los datos.

In [5]:
df_comparendos_electronicos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 339616 entries, 0 to 339615
Data columns (total 9 columns):
 #   Column                       Non-Null Count   Dtype 
---  ------                       --------------   ----- 
 0   fecha_comparendo             339616 non-null  object
 1   COD_INFRACCION               339616 non-null  object
 2   DESC_INFRACCION              339052 non-null  object
 3   TIPO_INFRACCION              339616 non-null  object
 4   SERVICIO_VEHICULO_INFRACTOR  339616 non-null  object
 5   CLASE_VEHICULO_INFRACTOR     339616 non-null  object
 6   CANTIDAD_INFRACCIONES        339616 non-null  int64 
 7   Tipo Camara                  339616 non-null  object
 8   Camara_y_direccion           339616 non-null  object
dtypes: int64(1), object(8)
memory usage: 23.3+ MB


El DataFrame `df_comparendos_electronicos` contiene **339,616 registros** (desde el índice 0 hasta el 339,615) y está compuesto por **9 columnas**. Las variables son mayoritariamente de tipo `object` (categóricas o texto), con una única columna numérica (`CANTIDAD_INFRACCIONES`) de tipo `int64`.


**Hallazgos:**

- **Volumen de datos**: Se cuenta con una base de datos robusta de más de 339 mil filas, lo que permite realizar análisis estadísticos representativos.
- **Valores nulos**: La columna `DESC_INFRACCION` presenta **564 valores nulos** (339,616 - 339,052), lo que indica que en algunos registros no se especificó la descripción de la infracción. El resto de las columnas están completas al 100%.
- **Tipos de datos**: Las columna de fechas (`fecha_comparendo`) está almacenada como `object`, por lo que se procederá a convertirlas a tipo `datetime` para facilitar análisis temporales.

### Análisis de la Hora en las Fechas

Se extraen y muestran las horas únicas presentes en la columna `fecha_comparendo`, con el fin de confirmar que todos los registros contienen la misma hora (12:00:00 AM) y justificar así la normalización realizada posteriormente.

In [6]:
print(f"Horas únicas: {df_comparendos_electronicos['fecha_comparendo'].str.split().str[3].unique()}")

Horas únicas: ['12:00:00']


### Conversión de las Fechas a Formato datetime

Se convierte la columna `fecha_comparendo` al tipo `datetime` utilizando el formato especificado (`'%Y %b %d %I:%M:%S %p'`), normalizando la hora a medianoche con el método `.dt.normalize()` para eliminar la información horaria y trabajar únicamente con fechas, dado que todos los registros contienen la hora de 12:00:00 AM. Posteriormente, se imprime el tipo de dato resultante para verificar la correcta conversión.

In [7]:
df_comparendos_electronicos['fecha_comparendo'] = pd.to_datetime(df_comparendos_electronicos['fecha_comparendo'], format='%Y %b %d %I:%M:%S %p').dt.normalize()

print(df_comparendos_electronicos['fecha_comparendo'].dtype)

datetime64[ns]


## Análisis de los Valores Nulos

Se construye una tabla que resume la cantidad y el porcentaje de valores nulos para cada columna del DataFrame, ordenando los resultados de mayor a menor porcentaje de nulos. Aunque previamente con el método `.info()` se identificó que la única columna con valores nulos era `DESC_INFRACCION`, este análisis se realiza para todas las variables con el fin de tener constancia y seguridad de la completitud del Dataset, confirmando nuevamente que no existen valores faltantes en otras columnas.

In [8]:
tabla_nulos = pd.DataFrame({
    'Nulos': df_comparendos_electronicos.isnull().sum(),
    '% Nulos': (df_comparendos_electronicos.isnull().sum() / len(df_comparendos_electronicos) * 100).round(4)
}).sort_values('% Nulos', ascending=False)

display(tabla_nulos)

,Nulos,% Nulos
DESC_INFRACCION,564,0.1661
fecha_comparendo,0,0.0000
COD_INFRACCION,0,0.0000
TIPO_INFRACCION,0,0.0000
SERVICIO_VEHICULO_INFRACTOR,0,0.0000
CLASE_VEHICULO_INFRACTOR,0,0.0000
CANTIDAD_INFRACCIONES,0,0.0000
Tipo Camara,0,0.0000
Camara_y_direccion,0,0.0000


Los resultados confirman que **únicamente la columna `DESC_INFRACCION` presenta valores nulos**, con un total de **564 registros** que equivalen al **0.17%**. El resto de las ocho columnas restantes no contienen ningún valor nulo, lo que representa una tasa de completitud del **99.83%** en el dataset.


**Hallazgos:**

- **Bajo porcentaje de nulos**: Con solo el 0.17% de datos faltantes en una sola columna, la calidad general de los datos es excelente, lo que minimiza la necesidad de aplicar técnicas de imputación complejas.
- **Impacto en el análisis**: La columna `DESC_INFRACCION` contiene la descripción textual de la infracción. Dado que existe la columna `COD_INFRACCION` (completamente poblada) que identifica cada infracción con un código único, los valores nulos en la descripción no representan una pérdida crítica de información, ya que el código puede utilizarse para obtener la descripción.

### Análisis de los Códigos de Infracción Asociados a Valores Nulos

Se filtran los registros que presentan valores nulos en la columna `DESC_INFRACCION` y se muestra la frecuencia de los códigos de infracción correspondientes, con el objetivo de identificar si los valores faltantes están concentrados en un código específico o si se distribuyen entre múltiples infracciones.

In [9]:
nulos_desc_infraccion = df_comparendos_electronicos[df_comparendos_electronicos['DESC_INFRACCION'].isna()]

print(nulos_desc_infraccion['COD_INFRACCION'].value_counts())

COD_INFRACCION
C14    564
Name: count, dtype: int64


### Análisis de los Tipos de Cámara Asociados a Valores Nulos

Se examina la frecuencia de los tipos de cámara asociados a los registros con valores nulos en `DESC_INFRACCION`, con el fin de identificar si existe una relación entre el tipo de cámara y la ausencia de descripción.

In [10]:
print(nulos_desc_infraccion['Tipo Camara'].value_counts())

Tipo Camara
Carril Bus    564
Name: count, dtype: int64


### Visualización de Registros con Valores Nulos

Posteriormente, se muestran las primeras filas de los registros filtrados que corresponden al código de infracción C14 y que carecen de descripción, con el propósito de examinar el contexto.

In [11]:
nulos_desc_infraccion[nulos_desc_infraccion['COD_INFRACCION'] == 'C14'].head()

,fecha_comparendo,COD_INFRACCION,DESC_INFRACCION,TIPO_INFRACCION,SERVICIO_VEHICULO_INFRACTOR,CLASE_VEHICULO_INFRACTOR,CANTIDAD_INFRACCIONES,Tipo Camara,Camara_y_direccion
332474,2025-10-21,C14,NaN,TRÁNSITO,PARTICULAR,AUTOMOVIL,1,Carril Bus,CALLE 45 CON CARRERA 1E SENTIDO NORTE-SUR
332475,2025-10-21,C14,NaN,TRÁNSITO,PUBLICO,AUTOMOVIL,1,Carril Bus,CALLE 45 CON CARRERA 1E SENTIDO NORTE-SUR
332476,2025-10-21,C14,NaN,TRÁNSITO,PUBLICO,AUTOMOVIL,2,Carril Bus,CALLE 45 CON CARRERA 1E SENTIDO SUR-NORTE
332477,2025-10-21,C14,NaN,TRÁNSITO,PUBLICO,CAMIONETA,1,Carril Bus,CALLE 45 CON CARRERA 1E SENTIDO SUR-NORTE
332478,2025-10-21,C14,NaN,TRÁNSITO,PARTICULAR,AUTOMOVIL,1,Carril Bus,CALLE 70 CON CARRERA 46 SENTIDO NORTE-SUR


**Explicación**:

La ausencia de descripción para el código C14 se explica porque este tipo de infracción comenzó a operar a partir del **17 de octubre de 2025** en la ciudad de **Barranquilla**, específicamente mediante el sistema de cámaras tipo **Carril Bus**. Es por esta razón que los primeros registros en la base de datos aparecen a partir del **21 de octubre de 2025**, coincidiendo con el inicio de operación de este sistema de vigilancia.

**Fuente:** https://barranquilla.gov.co/transito/desde-el-9-de-octubre-camaras-de-fotodeteccion-en-carriles-exclusivos-de-transmetro

---

**Hallazgos**:

- **Novedad del código**: Al tratarse de una infracción recientemente implementada la descripción asociada al código C14 no ha sido cargada en la base de datos durante el período inicial de operación.
- **Consistencia del patrón**: Todos los registros con código C14 comparten características homogéneas (cámaras Carril Bus, fechas posteriores al inicio de operación), lo que confirma que los valores nulos responden a un fenómeno sistemático relacionado con la incorporación de este nuevo tipo de infracción.


Los valores nulos en `DESC_INFRACCION` para el código C14 no representan una inconsistencia en los datos, sino la incorporación de un nuevo tipo de infracción cuya descripción aún no había sido registrada en el sistema durante el período inicial de operación de las cámaras Carril Bus en Barranquilla.

### Corrección de Valores Nulos

Se rellenan los valores nulos de la columna `DESC_INFRACCION` con la descripción correspondiente al código C14, obtenida del sitio web oficial del Tránsito del Atlántico. Según esta fuente, la infracción C14 corresponde a **"Transitar por sitios restringidos o en horas prohibidas por la autoridad competente"**. Esta corrección se aplica exclusivamente a los 564 registros que presentaban valores nulos, asociados a las nuevas cámaras tipo Carril Bus implementadas en Barranquilla a partir del 17 de octubre de 2025.

**Fuente:** https://transitodelatlantico.gov.co/valor-de-multas-de-transito/

In [12]:
df_comparendos_electronicos['DESC_INFRACCION'] = df_comparendos_electronicos['DESC_INFRACCION'].fillna("Transitar por sitios restringidos o en horas prohibidas por la autoridad competente")  

print(f"Total de valores nulos en el DataFrame: {df_comparendos_electronicos.isnull().sum().sum()}")

Total de valores nulos en el DataFrame: 0


## Análisis de Valores Únicos

Se construye una tabla que muestra para cada columna del DataFrame el tipo de dato y la cantidad de valores únicos presentes, ordenando los resultados de mayor a menor número de valores únicos. Este análisis permite identificar la diversidad de categorías en las variables y evaluar la variabilidad de los datos.

In [13]:
tabla_unicos = pd.DataFrame({
    'Variable': df_comparendos_electronicos.columns,
    'Tipo de Dato': df_comparendos_electronicos.dtypes.values,
    'Valores Únicos': df_comparendos_electronicos.nunique().values
}).sort_values('Valores Únicos', ascending=False).reset_index(drop=True)

display(tabla_unicos)

,Variable,Tipo de Dato,Valores Únicos
0,fecha_comparendo,datetime64[ns],2904
1,CANTIDAD_INFRACCIONES,int64,64
2,Camara_y_direccion,object,56
3,CLASE_VEHICULO_INFRACTOR,object,14
4,COD_INFRACCION,object,6
5,DESC_INFRACCION,object,6
6,SERVICIO_VEHICULO_INFRACTOR,object,4
7,Tipo Camara,object,3
8,TIPO_INFRACCION,object,1


**Variables con alta cardinalidad:**

- **`fecha_comparendo`**: Con 2,904 valores únicos, es la variable con mayor diversidad, lo que era de esperarse al tratarse de una serie temporal que abarca múltiples fechas. Esta alta cardinalidad permite realizar análisis de tendencias temporales con distintas agrupaciones.
- **`CANTIDAD_INFRACCIONES`**: Presenta 64 valores únicos, lo que indica que existen registros con diferentes números de infracciones por comparendo.


**Variables con cardinalidad media:**
- **`Camara_y_direccion`**: Con 56 valores únicos, muestra una variedad considerable de ubicaciones de cámaras, lo que permite análisis geográficos detallados de la ocurrencia de infracciones.
- **`CLASE_VEHICULO_INFRACTOR`**: Con 14 categorías, refleja una clasificación detallada de los tipos de vehículos involucrados en las infracciones.


**Variables con baja cardinalidad:**
- **`COD_INFRACCION`** y **`DESC_INFRACCION`**: Ambas presentan 6 valores únicos, lo que confirma la consistencia entre el código y la descripción después de la corrección de nulos. Las 6 categorías corresponden a los diferentes tipos de infracciones registradas en el sistema.
- **`SERVICIO_VEHICULO_INFRACTOR`**: Con 4 valores únicos, clasifica los vehículos en servicios.
- **`Tipo Camara`**: Con 3 valores únicos, identifica las diferentes tecnologías o tipos de cámaras utilizadas para la detección.


La variable `TIPO_INFRACCION` presenta **1 único valor**, lo que indica que en todos los registros el tipo de infracción es el mismo. Esto sugiere que la base de datos contiene únicamente infracciones de tránsito, sin presencia de otras categorías.

## Limitaciones Identificadas

- **Ausencia de hora real**: Todos los registros tienen la hora normalizada a `12:00:00 AM`, lo que impide análisis de franjas horarias dentro del día.

- **Sin identificación geográfica precisa**: La dirección de la cámara está en formato texto, no en coordenadas (latitud/longitud), lo que limita análisis espaciales avanzados.

- **Sin información del infractor**: No se incluyen datos como tipo de licencia, género, edad, o reincidencia individual, lo que restringe análisis demográficos.


**En resumen**, el dataset presenta una calidad excepcional en términos de completitud y consistencia, con una estructura adecuada para abordar análisis descriptivos, diagnósticos y predictivos relacionados con las infracciones de tránsito detectadas por cámaras electrónicas.